### PR TITLE
Feature/add find category controller endpoint

### PR DESCRIPTION
**Pull request by Segway-svg**

### CODE REVIEW

File: src/EventService.Models.Dto/Requests/Category/FindCategoriesFilter.cs [hunk 1]
```diff
@@ -0,0 +1,18 @@
+﻿using LT.DigitalOffice.EventService.Models.Dto.Enums;
+using LT.DigitalOffice.Kernel.Requests;
+using Microsoft.AspNetCore.Mvc;
+
+namespace LT.DigitalOffice.EventService.Models.Dto.Requests.Category;
+
+public record FindCategoriesFilter : BaseFindFilter
+{
+  [FromQuery(Name = "userFullNameIncludeSubstring")]
+  public string UserFullNameIncludeSubstring { get; set; }
```
Comment (SolovyovDmitry) [inline src/EventService.Models.Dto/Requests/Category/FindCategoriesFilter.cs]: too long, leave just NameIncludeSubstring

File: src/EventService.Models.Dto/Requests/Category/FindCategoriesFilter.cs [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+  [FromQuery(Name = "color")]
+  public CategoryColor? Color { get; set; }
+  
+  [FromQuery(Name = "isActive")]
+  public bool? IsActive { get; set; }
```
Comment (SolovyovDmitry) [inline src/EventService.Models.Dto/Requests/Category/FindCategoriesFilter.cs]: no need to make it nullable
usually we call it includeDeactivated

File: src/EventService/Controllers/CategoryController.cs [hunk 3]
```diff
@@ -19,5 +20,15 @@ public class CategoryController : ControllerBase
...
   }
+  
+  [HttpPost("find")]
+  public async Task<FindResultResponse<CategoryInfo>> FindCategoryFilter(
+    [FromServices] IFindCategoryCommand command,
+    [FromQuery] Guid categoryId,
+    [FromQuery] FindCategoriesFilter filter,
+    CancellationToken cancellationToken)
+  {
+    return await command.ExecuteAsync(eventId: categoryId, filter: filter, cancellationToken: cancellationToken);
```
Comment (SolovyovDmitry) [inline src/EventService/Controllers/CategoryController.cs]: separate lines
parameter is called 'eventId' and you pass 'categoryId'

File: src/EventService/Controllers/CategoryController.cs [hunk 4]
```diff
@@ -19,5 +20,15 @@ public class CategoryController : ControllerBase
   {
     return await command.ExecuteAsync(request);
   }
+  
+  [HttpPost("find")]
+  public async Task<FindResultResponse<CategoryInfo>> FindCategoryFilter(
+    [FromServices] IFindCategoryCommand command,
+    [FromQuery] Guid categoryId,
+    [FromQuery] FindCategoriesFilter filter,
+    CancellationToken cancellationToken)
```
Comment (SolovyovDmitry) [inline src/EventService/Controllers/CategoryController.cs]: can be just 'ct'

File: src/EventService.Models.Dto/Models/CategoryInfo.cs [hunk 5]
```diff
@@ -0,0 +1,13 @@
+﻿using System;
+using LT.DigitalOffice.EventService.Models.Dto.Enums;
+
+namespace LT.DigitalOffice.EventService.Models.Dto.Models;
+
+public class CategoryInfo
+{
+  public Guid Id { get; set; }
+  public string Name { get; set; }
+  public CategoryColor Color { get; set; }
```
Comment (SolovyovDmitry) [inline src/EventService.Models.Dto/Models/CategoryInfo.cs]: Can category not have color?

File: src/EventService.Business/Commands/Category/FindCategoryCommand.cs [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+using LT.DigitalOffice.EventService.Models.Dto.Requests.Category;
+using LT.DigitalOffice.Kernel.FluentValidationExtensions;
+using LT.DigitalOffice.Kernel.Helpers.Interfaces;
+using LT.DigitalOffice.Kernel.Responses;
+using LT.DigitalOffice.Kernel.Validators.Interfaces;
+using Microsoft.AspNetCore.Http;
+
+namespace LT.DigitalOffice.EventService.Business.Commands.Category;
+
+public class FindCategoryCommand : IFindCategoryCommand
```
Comment (SolovyovDmitry) [inline src/EventService.Business/Commands/Category/FindCategoryCommand.cs]: Categories

File: src/EventService.Business/Commands/Category/Interfaces/IFindCategoryCommand.cs [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+using System.Threading.Tasks;
+using LT.DigitalOffice.EventService.Models.Dto.Models;
+using LT.DigitalOffice.EventService.Models.Dto.Requests.Category;
+using LT.DigitalOffice.Kernel.Attributes;
+using LT.DigitalOffice.Kernel.Responses;
+
+namespace LT.DigitalOffice.EventService.Business.Commands.Category.Interfaces;
+
+[AutoInject]
+public interface IFindCategoryCommand
```
Comment (SolovyovDmitry) [inline src/EventService.Business/Commands/Category/Interfaces/IFindCategoryCommand.cs]: Categories

File: src/EventService.Data/CategoryRepository.cs [hunk 8]
```diff
@@ -3,16 +3,42 @@
...
 using LT.DigitalOffice.EventService.Data.Provider;
+using LT.DigitalOffice.EventService.Models.Dto.Requests.Category;
 
 namespace LT.DigitalOffice.EventService.Data;
 
 public class CategoryRepository : ICategoryRepository
 {
   private readonly IDataProvider _provider;
 
+  private async Task<IQueryable<DbCategory>> CreateFindPredicates(
```
Comment (SolovyovDmitry) [inline src/EventService.Data/CategoryRepository.cs]: you have no async code here so async can be removed

File: src/EventService.Data/CategoryRepository.cs [hunk 9]
```diff
@@ -3,16 +3,42 @@
...
+  FindCategoriesFilter filter,
+  Guid categoryId)
```
Comment (SolovyovDmitry) [inline src/EventService.Data/CategoryRepository.cs]: there should be no categoryId in find command. Id required for Get. For now you can find only 0 or 1 categories.

File: src/EventService/Controllers/CategoryController.cs [hunk 10]
```diff
@@ -19,5 +20,15 @@ public class CategoryController : ControllerBase
   {
     return await command.ExecuteAsync(request);
   }
+  
+  [HttpPost("find")]
+  public async Task<FindResultResponse<CategoryInfo>> FindCategoryFilter(
+    [FromServices] IFindCategoryCommand command,
+    [FromQuery] Guid categoryId,
```
Comment (SolovyovDmitry) [inline src/EventService/Controllers/CategoryController.cs]: remove from everywhere related to find

**Review by SolovyovDmitry — CHANGES_REQUESTED** [review]

File: src/EventService.Models.Dto/Models/CategoryInfo.cs (hunk 5 above)
Comment (Segway-svg) [reply src/EventService.Models.Dto/Models/CategoryInfo.cs]: Honestly, not sure. I think the category can be without color. You suggest to use "CategoryColor?"?

Comment (SolovyovDmitry) [reply src/EventService.Models.Dto/Models/CategoryInfo.cs]: yep

File: src/EventService/Controllers/CategoryController.cs [hunk 11]
```diff
@@ -19,5 +20,14 @@ public class CategoryController : ControllerBase
   {
     return await command.ExecuteAsync(request);
   }
+  
+  [HttpPost("find")]
```
Comment (liza459) [inline src/EventService/Controllers/CategoryController.cs]: `[HttpGet("find")]`

File: src/EventService/Controllers/CategoryController.cs [hunk 12]
```diff
@@ -19,5 +20,14 @@ public class CategoryController : ControllerBase
...
+  public async Task<FindResultResponse<CategoryInfo>> FindCategoryFilter(
```
Comment (liza459) [inline src/EventService/Controllers/CategoryController.cs:25]: Maybe this method would be better called **FindAsync**

File: src/EventService.Models.Dto/Requests/Category/FindCategoriesFilter.cs [hunk 13]
```diff
@@ -0,0 +1,18 @@
...
+public record FindCategoriesFilter : BaseFindFilter
+{
+  [FromQuery(Name = "nameIncludeSubstring")]
+  public string NameIncludeSubstring { get; set; }
+
+  [FromQuery(Name = "color")]
+  public CategoryColor? Color { get; set; }
+  
+  [FromQuery(Name = "includeDeactivated")]
+  public bool IncludeDeactivated { get; set; }
```
Comment (liza459) [inline src/EventService.Models.Dto/Requests/Category/FindCategoriesFilter.cs]: According to the terms of reference, categories are not archived.
The user should not have access to records that have IsActive = false
Mb add 
```
[FromQuery(Name = "isAscendingSort")]
 public bool? IsAscendingSort { get; set; }

```

File: src/EventService.Models.Dto/Models/CategoryInfo.cs [hunk 14]
```diff
@@ -0,0 +1,13 @@
...
+using LT.DigitalOffice.EventService.Models.Dto.Enums;
+
+namespace LT.DigitalOffice.EventService.Models.Dto.Models;
+
+public class CategoryInfo
+{
+  public Guid Id { get; set; }
+  public string Name { get; set; }
+  public CategoryColor? Color { get; set; }
+  public bool IsActive { get; set; }
```
Comment (liza459) [inline src/EventService.Models.Dto/Models/CategoryInfo.cs]: Remove IsActive

File: src/EventService.Mappers/Models/Interface/ICategoryInfoMapper.cs [hunk 15]
```diff
@@ -0,0 +1,12 @@
+using LT.DigitalOffice.EventService.Models.Db;
+using LT.DigitalOffice.EventService.Models.Dto.Models;
+using LT.DigitalOffice.Kernel.Attributes;
+
+namespace LT.DigitalOffice.EventService.Mappers.Models.Interface
```
Comment (liza459) [inline src/EventService.Mappers/Models/Interface/ICategoryInfoMapper.cs]: **"; "** at the end of the line

File: src/EventService.Mappers/Models/CategoryInfoMapper.cs [hunk 16]
```diff
@@ -0,0 +1,20 @@
...
+using LT.DigitalOffice.EventService.Models.Db;
+using LT.DigitalOffice.EventService.Models.Dto.Models;
+
+namespace LT.DigitalOffice.EventService.Mappers.Models
+{
+  public class CategoryInfoMapper : ICategoryInfoMapper
+  {
+    public CategoryInfo Map(DbCategory dbCategory)
+    {
+      return new CategoryInfo
```
Comment (liza459) [inline src/EventService.Mappers/Models/CategoryInfoMapper.cs]: If nothing is found and dbCategory is null

File: src/EventService.Mappers/Models/CategoryInfoMapper.cs [hunk 17]
```diff
@@ -0,0 +1,20 @@
+using LT.DigitalOffice.EventService.Mappers.Models.Interface;
+using LT.DigitalOffice.EventService.Models.Db;
+using LT.DigitalOffice.EventService.Models.Dto.Models;
+
+namespace LT.DigitalOffice.EventService.Mappers.Models
```
Comment (liza459) [inline src/EventService.Mappers/Models/CategoryInfoMapper.cs]: "; " at the end of the line

File: src/EventService.Data/CategoryRepository.cs [hunk 18]
```diff
@@ -3,16 +3,41 @@
...
 namespace LT.DigitalOffice.EventService.Data;
 
 public class CategoryRepository : ICategoryRepository
 {
   private readonly IDataProvider _provider;
 
+  private IQueryable<DbCategory> CreateFindPredicates(
+  FindCategoriesFilter filter)
+  {
+    IQueryable<DbCategory> dbCategories = _provider.Categories.AsNoTracking();
```
Comment (liza459) [inline src/EventService.Data/CategoryRepository.cs]: Here you can get all active records.

File: src/EventService.Data/CategoryRepository.cs [hunk 19]
```diff
@@ -3,16 +3,41 @@
...
+    {
+      dbCategories = dbCategories.Where(c => c.Color == filter.Color);
+    }
+
+    if (!filter.IncludeDeactivated)
+    {
+      dbCategories = dbCategories.Where(c => c.IsActive);
+    }
+
+    return dbCategories;
```
Comment (liza459) [inline src/EventService.Data/CategoryRepository.cs:40]: IQueryable will no longer be used. It is possible to materialize DbCategory into a list

File: src/EventService.Business/Commands/Category/FindCategoriesCommand.cs [hunk 20]
```diff
@@ -0,0 +1,61 @@
...
+  {
+    if (!_filterValidator.ValidateCustom(filter, out List<string> errors))
+    {
+      return _responseCreator.CreateFailureFindResponse<CategoryInfo>(
+        HttpStatusCode.BadRequest, errors);
+    }
+    
+    (List<DbCategory> dbCategories, int totalCount) = await _categoryRepository.FindAsync(filter, cancellationToken);
+
+    _contextAccessor.HttpContext.Response.StatusCode = (int)HttpStatusCode.OK;
```
Comment (liza459) [inline src/EventService.Business/Commands/Category/FindCategoriesCommand.cs]: This line can be removed. Status 200 will return anyway

File: src/EventService/Controllers/CategoryController.cs (hunk 12 above)
Comment (Segway-svg) [reply src/EventService/Controllers/CategoryController.cs:25]: We agreed with curators on this name

Comment (liza459) [reply src/EventService/Controllers/CategoryController.cs:25]: Ok

File: src/EventService.Data/CategoryRepository.cs (hunk 19 above)
Comment (SolovyovDmitry) [reply src/EventService.Data/CategoryRepository.cs:40]: can't agree. It's better to keep it until the final enumeration (return in FindAsync())

File: src/EventService/Controllers/CategoryController.cs (hunk 11 above)
Comment (SolovyovDmitry) [inline src/EventService/Controllers/CategoryController.cs]: Post?

**Review by SolovyovDmitry — CHANGES_REQUESTED** [review]
Next time before asking for review at least test your own changes. Since find endpoint has POST method, it's obvious that you haven't

File: src/EventService/Controllers/CategoryController.cs (hunk 11 above)
Comment (Segway-svg) [reply src/EventService/Controllers/CategoryController.cs]: Yeah, Liza has already pointed out this mistake. I will take it into account

File: src/EventService.Business/Commands/Category/FindCategoriesCommand.cs [hunk 21]
```diff
@@ -0,0 +1,59 @@
...
+    _mapper = mapper;
+    _responseCreator = responseCreator;
+  }
+  
+  
+  public async Task<FindResultResponse<CategoryInfo>> ExecuteAsync(
+    FindCategoriesFilter filter, 
+    CancellationToken cancellationToken = default)
+  {
+    if (!_filterValidator.ValidateCustom(filter, out List<string> errors))
```
Comment (SolovyovDmitry) [inline src/EventService.Business/Commands/Category/FindCategoriesCommand.cs]: We don't use it anymore this validation is inside base type of find filters

File: src/EventService.Data/CategoryRepository.cs [hunk 22]
```diff
@@ -3,16 +3,43 @@
...
+    if (filter.Color.HasValue)
+    {
+      dbCategories = dbCategories.Where(c => c.Color == filter.Color);
+    }
+    
+    if (filter.IsAscendingSort.HasValue)
+    {
+      dbCategories = filter.IsAscendingSort.Value
+        ? dbCategories.OrderBy(c => c.Id)
+        : dbCategories.OrderByDescending(c => c.Id);
```
Comment (liza459) [inline src/EventService.Data/CategoryRepository.cs]: sort by name, not id

File: src/EventService.Business/Commands/Category/FindCategoriesCommand.cs [hunk 23]
```diff
@@ -0,0 +1,51 @@
...
+    _filterValidator = filterValidator;
+    _contextAccessor = contextAccessor;
+    _mapper = mapper;
+    _responseCreator = responseCreator;
+  }
+  
+  
+  public async Task<FindResultResponse<CategoryInfo>> ExecuteAsync(
+    FindCategoriesFilter filter, 
+    CancellationToken cancellationToken = default)
```
Comment (liza459) [inline src/EventService.Business/Commands/Category/FindCategoriesCommand.cs:30]: default ? Why?

File: src/EventService.Business/Commands/Category/FindCategoriesCommand.cs [hunk 24]
```diff
@@ -0,0 +1,51 @@
...
+
+  public FindCategoriesCommand(
+  ICategoryRepository categoryRepository,
+  IBaseFindFilterValidator filterValidator,
+  IHttpContextAccessor contextAccessor,
+  ICategoryInfoMapper mapper,
+  IResponseCreator responseCreator)
+  {
+    _categoryRepository = categoryRepository;
+    _filterValidator = filterValidator;
```
Comment (liza459) [inline src/EventService.Business/Commands/Category/FindCategoriesCommand.cs:33]: remove unused variables

File: src/EventService.Business/Commands/Category/FindCategoriesCommand.cs (hunk 23 above)
Comment (SolovyovDmitry) [reply src/EventService.Business/Commands/Category/FindCategoriesCommand.cs:30]: that's ok. It gives opportunity not to pass ct

File: src/EventService.Mappers/Models/CategoryInfoMapper.cs [hunk 25]
```diff
@@ -0,0 +1,20 @@
...
+{
+  public CategoryInfo Map(DbCategory dbCategory)
+  {
+    return dbCategory is null
+      ? null
+      : new CategoryInfo 
+      {
+      Id = dbCategory.Id, 
+      Name = dbCategory.Name, 
+      Color = dbCategory.Color
```
Comment (SolovyovDmitry) [inline src/EventService.Mappers/Models/CategoryInfoMapper.cs]: tab

**Review by SolovyovDmitry — APPROVED** [review]

**Review by TheDioBrando — APPROVED** [review]

File: src/EventService.Business/Commands/Category/FindCategoriesCommand.cs [hunk 26]
```diff
@@ -0,0 +1,51 @@
...
+public class FindCategoriesCommand : IFindCategoriesCommand
+{
+  private readonly ICategoryRepository _categoryRepository;
+  private readonly IBaseFindFilterValidator _filterValidator;
+  private readonly IHttpContextAccessor _contextAccessor;
+  private readonly ICategoryInfoMapper _mapper;
+  private readonly IResponseCreator _responseCreator;
+
+  public FindCategoriesCommand(
+  ICategoryRepository categoryRepository,
```
Comment (VLiamin) [inline src/EventService.Business/Commands/Category/FindCategoriesCommand.cs]: tab?

**Review by VLiamin — APPROVED** [review]

File: src/EventService.Business/Commands/Category/FindCategoriesCommand.cs [hunk 27]
```diff
@@ -0,0 +1,51 @@
...
+  IBaseFindFilterValidator filterValidator,
+  IHttpContextAccessor contextAccessor,
+  ICategoryInfoMapper mapper,
+  IResponseCreator responseCreator)
+  {
+    _categoryRepository = categoryRepository;
+    _filterValidator = filterValidator;
+    _contextAccessor = contextAccessor;
+    _mapper = mapper;
+    _responseCreator = responseCreator;
```
Comment (liza459) [inline src/EventService.Business/Commands/Category/FindCategoriesCommand.cs]: filterValidator, contextAccessor, responseCreator - you don't call it anywhere.

File: src/EventService.Data/Interfaces/ICategoryRepository.cs [hunk 28]
```diff
@@ -11,4 +13,7 @@ public interface ICategoryRepository
 {
   bool DoesExistAllAsync(List<Guid> categoriesIds);
   Task<Guid?> CreateAsync(DbCategory dbCategory);
+  Task<(List<DbCategory> dbCategories, int totalCount)> FindAsync(
+    FindCategoriesFilter filter, 
+    CancellationToken cancellationToken);
```
Comment (liza459) [inline src/EventService.Data/Interfaces/ICategoryRepository.cs]: May be `cancellationToken = default` ?

**Review by liza459 — APPROVED** [review]

**Review by ndoriomedov — APPROVED** [review]

**Pull request merged**